cli/pkg/kctrl/cmd/package/release/helm_openapi_schema_gen.go

```diff
@@ -0,0 +1,362 @@
+		return nil, fmt.Errorf("values file must resolve to a map (was %d)", document.Content[0].Kind)
+	}
+	openAPIProperties, err := h.calculateProperties(nil, document.Content[0])
+	if err != nil {
+		return nil, err
+	}
+
+	bytes, err := yaml2.Marshal(h.convertToYAML(openAPIProperties))
+	if err != nil {
+		return nil, err
+	}
+	jsonEncodedBytes, err := yaml.YAMLToJSON(bytes)
+	if err != nil {
+		return nil, err
+	}
+	return &kcdatav1alpha1.ValuesSchema{
+		OpenAPIv3: runtime.RawExtension{Raw: jsonEncodedBytes},
+	}, nil
+}
+
+func (h HelmValuesSchemaGen) readValuesFile() ([]byte, error) {
+	fileInfo, err := os.Stat(h.dirPath)
+	if err != nil {
+		return nil, err
+	}
+	if !fileInfo.IsDir() {
+		return nil, fmt.Errorf("expected %s to be directory", h.dirPath)
+	}
+
+	fileData, err := os.ReadFile(filepath.Join(h.dirPath, "values.yaml"))
+	if err != nil {
+		// It is possible that helm chart doesn't have values.yml file.
+		if os.IsNotExist(err){
+			return nil, nil
+		}
+		return nil, err
+	}
+	return fileData, nil
+}
+
+func (h HelmValuesSchemaGen) convertToYAML(val interface{}) interface{} {
+	switch typedVal := val.(type) {
+	case *Map:
+		result := yaml2.MapSlice{}
+		typedVal.Iterate(func(k, v interface{}) {
+			result = append(result, yaml2.MapItem{
+				Key:   k,
+				Value: h.convertToYAML(v),
+			})
+		})
+		return result
+	default:
+		return val
+	}
+}
+
+func (h HelmValuesSchemaGen) calculateProperties(key *yaml3.Node, value *yaml3.Node) (*Map, error) {
+	switch value.Kind {
+	case yaml3.MappingNode:
+		var apiKeys openAPIKeys
+		description, isPresent := h.getDescriptionFromNode(key)
+		if isPresent {
+			apiKeys = append(apiKeys, &MapItem{Key: descriptionKey, Value: description})
+		}
+		apiKeys = append(apiKeys, &MapItem{Key: typeKey, Value: objectVal})
+
+		var properties openAPIKeys
+		for i := 0; i < len(value.Content); i += 2 {
+			k := value.Content[i]
+			v := value.Content[i+1]
+			calculatedProperties, err := h.calculateProperties(k, v)
+			if err != nil {
+				return nil, err
+			}
+			valueItems := calculatedProperties.Items[0]
+			properties = append(properties, &MapItem{Key: valueItems.Key, Value: valueItems.Value})
+		}
+		if len(value.Content) == 0 {
+			apiKeys = append(apiKeys, &MapItem{Key: defaultKey, Value: "{}"})
+		} else {
+			apiKeys = append(apiKeys, &MapItem{Key: propertiesKey, Value: &Map{Items: properties}})
+		}
+
+		sort.Sort(apiKeys)
+		if key == nil {
+			return &Map{Items: apiKeys}, nil
+		}
+		return &Map{Items: []*MapItem{&MapItem{Key: key.Value, Value: &Map{Items: apiKeys}}}}, nil
+	case yaml3.SequenceNode:
+		var defaultVals []interface{}
+		var apiKeys openAPIKeys
+		apiKeys = append(apiKeys, &MapItem{Key: typeKey, Value: arrayVal})
+		description, isPresent := h.getDescriptionFromNode(key)
+		if isPresent {
+			apiKeys = append(apiKeys, &MapItem{Key: descriptionKey, Value: description})
+		}
+
+		if len(value.Content) != 0 {
+			properties := &Map{[]*MapItem{}}
+			for _, v := range value.Content {
```

Comment:
Similar to above, this for loop can be out of the if condition.

Reply:
Even if I move it out, there is still 1 if condition needs to be added. I think this way is cleaner.